schema/shared/src/main/scala-3/zio/blocks/schema/DerivedOptics.scala

```diff
@@ -0,0 +1,318 @@
+  )(using q: Quotes): Expr[Any] = {
+    import q.reflect.*
+
+    val symCast = sym.asInstanceOf[Symbol]
+    val tpeCast = tpe.asInstanceOf[TypeRepr]
+
+    val fields = symCast.caseFields
+
+    // Build the structural refinement type
+    // Start with OpticsHolder as the base type
+    var refinedType: TypeRepr = TypeRepr.of[OpticsHolder]
+
+    // Add a refinement for each field: def fieldName: Lens[S, FieldType]
+    for (field <- fields) {
+      val fieldType    = tpeCast.memberType(field).dealias
+      val lensType     = TypeRepr.of[Lens].appliedTo(List(tpeCast, fieldType))
+      val accessorName = if (prefixUnderscore) "_" + field.name else field.name
+      refinedType = Refinement(refinedType, accessorName, lensType)
+    }
+
+    // Get unique type string at compile time for the cache key
+    // We use show to ensure generic types like Box[Int] and Box[String] have different keys
+    val cacheKey: Expr[String]              = Expr(tpeCast.show + (if (prefixUnderscore) "_" else ""))
+    val prefixUnderscoreExpr: Expr[Boolean] = Expr(prefixUnderscore)
+
+    // Match the refined type and create the implementation
+    refinedType.asType match {
+      case '[t] =>
+        '{
+          getOrCreate(
+            $cacheKey, {
+              val record = $schema.reflect.asRecord.getOrElse(
+                throw new RuntimeException(s"Expected a record schema for ${$cacheKey}")
+              )
+              val members = record.fields.zipWithIndex.map { case (term, idx) =>
+                val lens = record
+                  .lensByIndex(idx)
+                  .getOrElse(
+                    throw new RuntimeException(s"Cannot find lens for field ${term.name}")
+                  )
+                val name = if ($prefixUnderscoreExpr) "_" + term.name else term.name
+                name -> lens
+              }.toMap
+              new OpticsHolder(members)
+            }
+          ).asInstanceOf[t]
+        }
+    }
+  }
+
+  private def buildSealedTraitOptics[S: Type](
+    schema: Expr[Schema[S]],
+    sym: Quotes#reflectModule#Symbol,
+    tpe: Quotes#reflectModule#TypeRepr,
+    prefixUnderscore: Boolean
+  )(using q: Quotes): Expr[Any] = {
+    import q.reflect.*
+
+    val symCast = sym.asInstanceOf[Symbol]
+    val tpeCast = tpe.asInstanceOf[TypeRepr]
+
+    val children = symCast.children
+
+    // Build the structural refinement type
+    var refinedType: TypeRepr = TypeRepr.of[OpticsHolder]
+
+    // Add a refinement for each child: def childName: Prism[S, ChildType]
+    for (child <- children) {
+      val childType = if (child.isType) {
+        // If the child has type parameters, try to apply the parent's type arguments
+        // This handles the common pattern: sealed trait T[A]; case class C[A](...) extends T[A]
+        // We check primaryConstructor paramSymss to detect type parameters
+        val hasTypeParams = child.primaryConstructor.paramSymss.headOption.exists(_.exists(_.isType))
```

Comment:
This needs some hand-written unit tests designed to force failure, because this seems not robust.